rust/agama-server/src/agama-web-server.rs

```diff
@@ -32,23 +60,213 @@ struct Cli {
+    // "GET ... HTTP/1.1" or "POST ... HTTP/1.1" or a similar line
+    buf[0] == 0x16u8 || buf[0] == 0x80u8
+}
+
+// build a SSL acceptor using a provided SSL certificate or generate a self-signed one
+fn create_ssl_acceptor(cert_file: &String, key_file: &String) -> SslAcceptor {
+    let mut tls_builder = SslAcceptor::mozilla_modern_v5(SslMethod::tls_server()).unwrap();
+
+    if cert_file.is_empty() && key_file.is_empty() {
+        let (cert, key) = agama_dbus_server::cert::create_certificate().unwrap();
+        tracing::info!("Generated self signed certificate: {:#?}", cert);
+        tls_builder.set_private_key(key.as_ref()).unwrap();
+        tls_builder.set_certificate(cert.as_ref()).unwrap();
+
+        // for debugging you might dump the certificate to a file:
+        // use std::io::Write;
+        // let mut cert_file = std::fs::File::create("agama_cert.pem").unwrap();
+        // let mut key_file = std::fs::File::create("agama_key.pem").unwrap();
+        // cert_file.write_all(cert.to_pem().unwrap().as_ref()).unwrap();
+        // key_file.write_all(key.private_key_to_pem_pkcs8().unwrap().as_ref()).unwrap();
+    } else {
+        tracing::info!("Loading PEM certificate: {}", cert_file);
+        tls_builder
+            .set_certificate_file(PathBuf::from(cert_file), SslFiletype::PEM)
+            .unwrap();
+
+        tracing::info!("Loading PEM key: {}", key_file);
+        tls_builder
+            .set_private_key_file(PathBuf::from(key_file), SslFiletype::PEM)
+            .unwrap();
+    }
+
+    // check that the key belongs to the certificate
+    tls_builder.check_private_key().unwrap();
+
+    tls_builder.build()
+}
+
+// build a response for the HTTP -> HTTPS redirection
+// returns 308 permanent redirect
+fn redirect_https(host: &str, uri: &hyper::Uri) -> Response<String> {
+    let builder = Response::builder()
+        // build the redirection target URL
+        .header("Location", format!("https://{}{}", host, uri))
+        .status(hyper::StatusCode::PERMANENT_REDIRECT);
+
+    builder.body(String::from("")).unwrap()
+}
+
+// build an error response for the HTTP -> HTTPS redirection when we cannot build
+// the redirect response from the original request
+// returns error 400
+fn redirect_error() -> Response<String> {
+    let builder = Response::builder().status(hyper::StatusCode::BAD_REQUEST);
+
+    let msg = "HTTP protocol is not allowed for external requests, please use HTTPS.\n";
+    builder.body(String::from(msg)).unwrap()
+}
+
+// build a router for the HTTP -> HTTPS redirection
+// if the redirection URL cannot be built from the original request it returns error 400
+// instead of the redirection
+fn https_redirect() -> Router {
+    // see https://docs.rs/axum/latest/axum/routing/struct.Router.html#example
+    let redirect_service = tower::service_fn(|req: axum::extract::Request| async move {
+        let request_host = req.headers().get("host");
+        match request_host {
```

Comment:
to avoid so much nesting you can use `if let syntax`

```
if let Some(host) = request_host {
  let host_string = host.to_str();
  if let Ok(host_str) = host_string {
     Ok(redirect_https(host_str, req.uri())
  } else {
    Ok(redirect_error())
      } else {
  Ok(redirect_error())

  }
```